packages/wrangler/src/api/startDevWorker/BundlerController.ts

```diff
@@ -1,37 +1,280 @@
+import type { EsbuildBundle } from "../../dev/use-esbuild";
+import type { EphemeralDirectory } from "../../paths";
 import type { ControllerEventMap } from "./BaseController";
 import type {
 	BundleCompleteEvent,
 	BundleStartEvent,
 	ConfigUpdateEvent,
 } from "./events";
+import type { StartDevWorkerOptions } from "./types";
 
 export type BundlerControllerEventMap = ControllerEventMap & {
 	bundleStart: [BundleStartEvent];
 	bundleComplete: [BundleCompleteEvent];
 };
 export class BundlerController extends Controller<BundlerControllerEventMap> {
-	// ******************
-	//   Event Handlers
-	// ******************
+	#currentBundle?: EsbuildBundle;
 
-	onConfigUpdate(_: ConfigUpdateEvent) {
-		notImplemented(this.onConfigUpdate.name, this.constructor.name);
+	#customBuildWatcher?: ReturnType<typeof watch>;
+
+	// Handle aborting in-flight custom builds as new ones come in from the filesystem watcher
+	// Note: we don't need this for non-custom builds since esbuild handles this internally with it's watch mode
+	#customBuildAborter = new AbortController();
+
+	async #runCustomBuild(config: StartDevWorkerOptions, filePath: string) {
+		// If a new custom build comes in, we need to cancel in-flight builds
+		this.#customBuildAborter.abort();
+		this.#customBuildAborter = new AbortController();
+
+		// Since `this.#customBuildAborter` will change as new builds are scheduled, store the specific AbortController that will be used for this build
+		const buildAborter = this.#customBuildAborter;
+		assert(config._entry);
+		const relativeFile =
+			path.relative(config._entry.directory, config._entry.file) || ".";
+		logger.log(`The file ${filePath} changed, restarting build...`);
+		this.emitBundleStartEvent({
+			type: "bundleStart",
+			config,
+		});
+		try {
+			await runCustomBuild(config._entry.file, relativeFile, {
+				cwd: config.build?.custom.workingDirectory,
+				command: config.build?.custom.command,
+			});
+			if (buildAborter.signal.aborted) {
+				return;
+			}
+			assert(this.#tmpDir);
+			assert(config._entry, "config._entry");
+			assert(config._additionalModules, "config._additionalModules");
+			assert(config.build?.moduleRules, "config.build?.moduleRules");
+			assert(config.build?.define, "config.build?.define");
+			if (!config.build?.bundle) {
+				// if we're not building, let's just copy the entry to the destination directory
+				const destinationDir = this.#tmpDir.path;
+				writeFileSync(
+					path.join(destinationDir, path.basename(config._entry.file)),
+					readFileSync(config._entry.file, "utf-8")
+				);
+			}
+
+			const entryDirectory = path.dirname(config._entry.file);
+			const moduleCollector = createModuleCollector({
+				wrangler1xLegacyModuleReferences: getWrangler1xLegacyModuleReferences(
+					entryDirectory,
+					config._entry.file
+				),
+				entry: config._entry,
+				// `moduleCollector` doesn't get used when `props.noBundle` is set, so
```

Comment:
There is no `props`:

```suggestion
				// `moduleCollector` doesn't get used when `noBundle` is set, so
```